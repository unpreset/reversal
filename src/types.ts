import type { CssNode } from 'css-tree'

export type Arrayable<T> = T | T[]
export type Objectable<T> = Record<string, T>
export interface Colors {
  [key: string]: Colors & { DEFAULT?: string } | string
}

export type Atomic = string
export type AtomicShort = string
export type AtomicComposed = [Atomic, AtomicShort?]
export type StaticPropAtomicMap = [string, Atomic | AtomicComposed]
export type DynamicPropAtomicMap = [RegExp, (match: RegExpMatchArray) => Atomic | AtomicComposed | undefined]
export type PropsAtomicMap = StaticPropAtomicMap | DynamicPropAtomicMap

// Parser types

export interface CssValueParsedMeta {
  /**
   * Value of the property or function
   */
  value: string | CssValueParsedMeta[]
  /**
   * Unit of the value
   */
  unit?: string
  /**
   * Function name
   */
  fname?: string
  /**
   * Type of the value
   */
  type: CssNode['type']
  /**
   * Raw string of the value
   */
  raw: string
}

export interface CssValueParsed {
  prop: string
  meta: CssValueParsedMeta[]
}

export interface TransfromOptions {
  /**
   * 是否开启简写模式
   *
   * @default false
   */
  shortify?: boolean
}
