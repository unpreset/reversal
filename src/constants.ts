export const EmptyKeyPlaceholder = '__EMPTY__'
