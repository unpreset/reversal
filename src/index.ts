import type { Rule, StyleSheet } from 'css-tree'
import { parse } from 'css-tree'
import MagicString from 'magic-string'

export * from './parser/declaration'

function blockSource(source: string) {
  if (source.includes('{') && source.includes('}')) {
    return source
  }
  return `{ ${source} }`
}

export function parseCSS(css: string) {
  const source = new MagicString(css)
  const ast = parse(blockSource(source.original)) as StyleSheet

  for (const node of ast.children) {
    if (node.type === 'Rule') {
      handleRuleNode(source, node)
    }
  }
}

function handleRuleNode(source: MagicString, node: Rule) {
//   console.dir(node)
//   const selector = node.prelude.children
}
