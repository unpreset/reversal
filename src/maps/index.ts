import { border } from './border'
import { margin } from './margin'
import { positions } from './position'

export const maps = [
  border,
  positions,
  margin,
].flat(1)
