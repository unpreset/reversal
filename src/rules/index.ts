import type { DynamicPropAtomicMap, StaticPropAtomicMap } from '../types'
import { border } from './border'
import { colors } from './color'
import { displays } from './display'
import { positions } from './position'
import { sizes } from './size'
import { spacing } from './spacing'

export const rules = [
  border,
  positions,
  spacing,
  colors,
  sizes,
  displays,
].flat(1)

export const staticRules: StaticPropAtomicMap[] = []
export const dynamicRules: DynamicPropAtomicMap[] = []

export const defaultRuleMeta = {
  separator: '-',
  valueProcessor: (v: string) => v,
}

for (const rule of rules) {
  if (typeof rule[0] === 'string') {
    staticRules.push(rule as StaticPropAtomicMap)
  }
  else {
    dynamicRules.push(rule as DynamicPropAtomicMap)
  }
}
