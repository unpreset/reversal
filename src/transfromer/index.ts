import { maps } from '../maps'
import { toArray } from '../utils'
import type { AtomicComposed, CssValueParsed, CssValueParsedMeta, DynamicPropAtomicMap, StaticPropAtomicMap, TransfromOptions } from '../types'

const atomicCache: Record<string, string[]> = {}

const nonTransfromPxProps = [
  'border',
]

/**
 * 将 CssValueParsedMeta[] 转换为 atomic css
 * @param meta CssValueParsedMeta[]
 * @returns
 */
export function transfrom(metas: CssValueParsedMeta[], options: TransfromOptions = {}): string[] {
  const atomics: string[] = []
}

export function transfromParsed(parsed: CssValueParsed, options: TransfromOptions = {}) {
  let key: string | undefined

  const {
    shortify = false,
  } = options
  const { prop, meta } = parsed

  for (const map of maps) {
    let atomics: AtomicComposed | undefined

    if (typeof map[0] === 'string') {
      if (prop === map[0]) {
        atomics = toArray((map as StaticPropAtomicMap)[1]) as AtomicComposed
      }
    }
    else {
      const match = prop.match(map[0])
      if (match) {
        const matched = (map as DynamicPropAtomicMap)[1](match)
        if (matched) {
          atomics = toArray(matched) as AtomicComposed
        }
      }
    }

    if (atomics) {
      if (shortify) {
        key = atomics[1] || atomics[0]
      }
      else {
        key = atomics[0]
      }
      break
    }
  }

  if (!key) {
    return
  }

  return analyzeMeta({
    prop,
    meta,
    key,
  })
}

function analyzeMeta(bundle: {
  meta: CssValueParsed['meta']
  key: string
  prop: string
}): string[] {
  const { meta, key, prop } = bundle
  const atomics: string[] = []

  for (const m of meta) {
    if (Array.isArray(m)) {

    }
    else {
      if (m.unit === 'px' && /^\d+$/.test(m.value as string)) {
        if (nonTransfromPxProps.some(p => prop.includes(p))) {
          atomics.push(`${key}-[${m.value}${m.unit}]`)
        }
        else {
          atomics.push(`${key}-${Number(m.value) / 4}`)
        }
      }
    }
  }

  return atomics
}
