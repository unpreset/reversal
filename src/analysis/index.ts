import type { UnoGenerator } from '@unocss/core'
import { EmptyKeyPlaceholder } from '../constants'
import { defaultRuleMeta, dynamicRules, staticRules } from '../rules'
import { guessColorType, isColorProp, transformColor } from '../transformers/color'
import { noop, toArray } from '../utils'
import type { AtomicComposed, AtomicMeta, Colors, CssValueParsed, CssValueParsedMeta, DynamicPropAtomicMap, ProcessorContext, StaticPropAtomicMap, TransfromOptions } from '../types'

const atomicCache: Record<string, string[]> = {}

const nonTransfromPxProps = [
  'border',
]

/**
 * 将 CssValueParsedMeta[] 转换为 atomic css
 * @param meta CssValueParsedMeta[]
 * @returns
 */
export function transfrom(metas: CssValueParsedMeta[], options: TransfromOptions = {}): string[] {

}

export function analyzePared(parsed: CssValueParsed, uno: UnoGenerator, options: TransfromOptions = {}) {
  const {
    shortify = false,
  } = options
  const { prop, meta } = parsed
  let { key, meta: ruleMeta } = findKeyMeta(prop, shortify)

  if (!key)
    return
  if (ruleMeta?.keyProcessor)
    key = ruleMeta.keyProcessor(key)

  return analyzeMeta({
    prop,
    meta,
    uno,
    key: key === EmptyKeyPlaceholder ? '' : key,
    ruleMeta,
  })
}

function analyzeMeta(bundle: {
  meta: CssValueParsed['meta']
  key: string
  prop: string
  uno: UnoGenerator
  ruleMeta?: AtomicMeta
}): string[] {
  const { meta, key, prop, uno, ruleMeta } = bundle
  const { separator, valueProcessor } = Object.assign({}, defaultRuleMeta, ruleMeta ?? {})
  const atomics: string[] = []

  for (const m of meta) {
    let token: string | undefined

    if (Array.isArray(m.value)) {
      if (m.type === 'Function') {
        if (m.fname === 'var') {
          token = m.value.map(v => (v.value as string).trim()).join('').replace(/^--/, '$')
        }
        else {
          token = `[${m.raw.replace(/\s/g, '_')}]`
        }
      }
    }
    else {
      if (m.unit === 'px' && /^\d+$/.test(m.value as string)) {
        if (nonTransfromPxProps.some(p => prop.includes(p))) {
          token = `${m.value}${m.unit}`
        }
        else {
          token = `${Number(m.value) / 4}`
        }
      }

      else if (isColorProp(prop) || m.type === 'Hash' || guessColorType(m.raw) != null) {
        const themeKey = transformColor(m.value as string, uno)
        if (themeKey) {
          token = themeKey
        }
        else {
          token = m.value
        }
      }

      else if (m.type === 'Identifier') {
        token = m.value
      }
      else if (m.type === 'Percentage') {
        if (m.value === '100') {
          token = 'full'
        }
        else {
          token = `${m.value}%`
        }
      }

      else {
        token = m.raw
      }
    }

    if (token) {
      token = valueProcessor(token, { uno })
      atomics.push(`${key}${separator}${token}`)
    }
  }

  return atomics
}

function findKeyMeta(prop: string, shortify = false) {
  let key: string | undefined
  let atomics: AtomicComposed | undefined
  let rule: StaticPropAtomicMap | DynamicPropAtomicMap | undefined
  let meta: AtomicMeta | undefined

  for (const r of staticRules) {
    if (prop === r[0]) {
      rule = r
      atomics = toArray((r as StaticPropAtomicMap)[1]) as AtomicComposed
      break
    }
  }

  if (!rule) {
    for (const r of dynamicRules) {
      const match = prop.match(r[0])
      if (match) {
        rule = r
        atomics = toArray(r[1](match)) as AtomicComposed
        break
      }
    }
  }

  if (atomics && rule) {
    if (shortify) {
      key = atomics[1] || atomics[0]
    }
    else {
      key = atomics[0]
    }
    meta = rule[2]
  }

  return { key, meta }
}
