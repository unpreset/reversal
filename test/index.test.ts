import { Magicolor } from '@magic-color/core'
import { createGenerator } from '@unocss/core'
import { parse } from 'css-tree'
import { describe, expect, it } from 'vitest'
import type { Declaration } from 'css-tree'
import { parseCSS, parseDeclarationNode } from '../src'
import { transfromParsed } from '../src/transfromer'

const source = `/**
 * Paste or drop some CSS here and explore
 * the syntax tree created by chosen parser.
 * Enjoy!
 */

@media screen and (min-width: 480px) {
    body {
        background-color: lightgreen;
    }
}

#main {
    border: 1px solid black;
}

ul li {
	padding: 5px;
  color: #fff;
}
`

const uno = createGenerator({
  theme: {
    colors: {
      white: '#fff',
      primary: '#007bff',
      foo: {
        DEFAULT: '#f00',
        100: '#f0f',
        200: '#0ff',
        300: 'rgb(0 0 0)',
      },
      bar: {
        foo: {
          test: '#f00',
        },
      },
    },
  },
})

describe('should', () => {
  function generateParsed(code: string) {
    const ast = parse(code, {
      context: 'declaration',
      positions: true,
    }) as Declaration
    return parseDeclarationNode(ast)
  }

  it('declarations', () => {
    expect(generateParsed('content: ""')).toMatchInlineSnapshot(`
      {
        "meta": [
          {
            "type": "String",
            "value": "",
          },
        ],
        "prop": "content",
      }
    `)

    expect(generateParsed('color: red')).toMatchInlineSnapshot(`
      {
        "meta": [
          {
            "type": "Identifier",
            "value": "red",
          },
        ],
        "prop": "color",
      }
    `)

    expect(generateParsed('border: 1px solid #eee')).toMatchInlineSnapshot(`
      {
        "meta": [
          {
            "type": "Dimension",
            "unit": "px",
            "value": "1",
          },
          {
            "type": "Identifier",
            "value": "solid",
          },
          {
            "type": "Hash",
            "value": "#eee",
          },
        ],
        "prop": "border",
      }
    `)

    expect(generateParsed('background-color: hsl(100% var(--foo) 100 / 1)')).toMatchInlineSnapshot(`
      {
        "meta": [
          {
            "fname": "hsl",
            "type": "Function",
            "value": [
              {
                "type": "Percentage",
                "value": "100%",
              },
              {
                "fname": "var",
                "type": "Function",
                "value": [
                  {
                    "type": "Identifier",
                    "value": "--foo",
                  },
                ],
              },
              {
                "type": "Number",
                "value": "100",
              },
              {
                "type": "Operator",
                "value": "/",
              },
              {
                "type": "Number",
                "value": "1",
              },
            ],
          },
        ],
        "prop": "background-color",
      }
    `)

    expect(generateParsed('margin-top: calc(10px + calc(var(--bar, 1)))')).toMatchInlineSnapshot(`
      {
        "meta": [
          {
            "fname": "calc",
            "type": "Function",
            "value": [
              {
                "type": "Dimension",
                "unit": "px",
                "value": "10",
              },
              {
                "type": "Operator",
                "value": " + ",
              },
              {
                "fname": "calc",
                "type": "Function",
                "value": [
                  {
                    "fname": "var",
                    "type": "Function",
                    "value": [
                      {
                        "type": "Identifier",
                        "value": "--bar",
                      },
                      {
                        "type": "Operator",
                        "value": ",",
                      },
                      {
                        "type": "Raw",
                        "value": " 1",
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
        "prop": "margin-top",
      }
    `)
  })

  it('transfromParsed', () => {
    expect(transfromParsed(
      generateParsed('border-top: 1px solid #eee')!,
      uno as any,
      { shortify: true },
    )).toMatchInlineSnapshot(`
      [
        "b-t-1px",
      ]
    `)

    expect(transfromParsed(
      generateParsed('margin: 12px')!,
      uno as any,
      { shortify: true },
    )).toMatchInlineSnapshot(`
      [
        "m-3",
      ]
    `)
  })

  it('color in theme', () => {
    expect(transfromParsed(
      generateParsed('color: foo')!,
      uno as any,
      { shortify: true },
    )).toMatchInlineSnapshot(`
      [
        "c-[foo]",
      ]
    `)
  })
})
