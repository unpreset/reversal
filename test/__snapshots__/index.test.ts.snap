// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyzePared > transfrom parsed 1`] = `
[
  {
    "css": "border-top: 1px solid #eee",
    "shortTokens": [
      "b-t-1px",
      "b-t-solid",
      "b-t-#eee",
    ],
    "tokens": [
      "border-t-1px",
      "border-t-solid",
      "border-t-#eee",
    ],
  },
  {
    "css": "border: 1px solid rgb(0 0 0)",
    "shortTokens": [
      "b-1px",
      "b-solid",
      "b-[rgb(0_0_0)]",
    ],
    "tokens": [
      "border-1px",
      "border-solid",
      "border-[rgb(0_0_0)]",
    ],
  },
  {
    "css": "margin: 12px",
    "shortTokens": [
      "m-3",
    ],
    "tokens": [
      "m-3",
    ],
  },
  {
    "css": "background-color: hsl(100% var(--foo) 100 / 1)",
    "shortTokens": [
      "bg-[hsl(100%_var(--foo)_100_/_1)]",
    ],
    "tokens": [
      "bg-[hsl(100%_var(--foo)_100_/_1)]",
    ],
  },
  {
    "css": "color: #eee",
    "shortTokens": [
      "c-#eee",
    ],
    "tokens": [
      "c-#eee",
    ],
  },
  {
    "css": "color: var(--bar)",
    "shortTokens": [
      "c-$bar",
    ],
    "tokens": [
      "c-$bar",
    ],
  },
  {
    "css": "padding: var(--bar, 1)",
    "shortTokens": [
      "p-$bar,1",
    ],
    "tokens": [
      "p-$bar,1",
    ],
  },
  {
    "css": "font-size: 16px",
    "shortTokens": [
      "text-4",
    ],
    "tokens": [
      "text-4",
    ],
  },
  {
    "css": "line-height: 1.5",
    "shortTokens": [
      "lh-1.5",
    ],
    "tokens": [
      "leading-1.5",
    ],
  },
  {
    "css": "text-align: center",
    "shortTokens": undefined,
    "tokens": undefined,
  },
  {
    "css": "display: flex",
    "shortTokens": [
      "flex",
    ],
    "tokens": [
      "flex",
    ],
  },
  {
    "css": "justify-content: space-between",
    "shortTokens": undefined,
    "tokens": undefined,
  },
  {
    "css": "align-items: center",
    "shortTokens": undefined,
    "tokens": undefined,
  },
  {
    "css": "width: 100%",
    "shortTokens": [
      "w-full",
    ],
    "tokens": [
      "w-full",
    ],
  },
  {
    "css": "height: 50vh",
    "shortTokens": [
      "h-50vh",
    ],
    "tokens": [
      "h-50vh",
    ],
  },
  {
    "css": "max-width: 1200px",
    "shortTokens": [
      "max-w-300",
    ],
    "tokens": [
      "max-w-300",
    ],
  },
  {
    "css": "min-height: 300px",
    "shortTokens": [
      "min-h-75",
    ],
    "tokens": [
      "min-h-75",
    ],
  },
  {
    "css": "padding-top: 20px",
    "shortTokens": [
      "p-t-5",
    ],
    "tokens": [
      "p-t-5",
    ],
  },
  {
    "css": "padding-right: 10px",
    "shortTokens": [
      "p-r-2.5",
    ],
    "tokens": [
      "p-r-2.5",
    ],
  },
  {
    "css": "padding-bottom: 20px",
    "shortTokens": [
      "p-b-5",
    ],
    "tokens": [
      "p-b-5",
    ],
  },
  {
    "css": "padding-left: 10px",
    "shortTokens": [
      "p-l-2.5",
    ],
    "tokens": [
      "p-l-2.5",
    ],
  },
  {
    "css": "margin-left: auto",
    "shortTokens": [
      "m-l-auto",
    ],
    "tokens": [
      "m-l-auto",
    ],
  },
  {
    "css": "margin-right: auto",
    "shortTokens": [
      "m-r-auto",
    ],
    "tokens": [
      "m-r-auto",
    ],
  },
  {
    "css": "border-radius: 5px",
    "shortTokens": [
      "rd-5px",
    ],
    "tokens": [
      "rounded-5px",
    ],
  },
  {
    "css": "border-top-left-radius: 5px",
    "shortTokens": [
      "rd-tl-5px",
    ],
    "tokens": [
      "rounded-tl-5px",
    ],
  },
  {
    "css": "box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1)",
    "shortTokens": [
      "shadow-0",
      "shadow-1",
      "shadow-2",
      "shadow-[rgba(0,_0,_0,_0.1)]",
    ],
    "tokens": [
      "shadow-0",
      "shadow-1",
      "shadow-2",
      "shadow-[rgba(0,_0,_0,_0.1)]",
    ],
  },
  {
    "css": "opacity: 0.8",
    "shortTokens": [
      "op-80",
    ],
    "tokens": [
      "opacity-80",
    ],
  },
  {
    "css": "transition: all 0.3s ease",
    "shortTokens": undefined,
    "tokens": undefined,
  },
  {
    "css": "transform: translateX(50%)",
    "shortTokens": undefined,
    "tokens": undefined,
  },
  {
    "css": "background: linear-gradient(to right, #ff7e5f, #feb47b)",
    "shortTokens": [
      "bg-[linear-gradient(to_right,_#ff7e5f,_#feb47b)]",
    ],
    "tokens": [
      "bg-[linear-gradient(to_right,_#ff7e5f,_#feb47b)]",
    ],
  },
  {
    "css": "overflow: hidden",
    "shortTokens": [
      "of-hidden",
    ],
    "tokens": [
      "overflow-hidden",
    ],
  },
  {
    "css": "z-index: 10",
    "shortTokens": [
      "z-10",
    ],
    "tokens": [
      "z-10",
    ],
  },
  {
    "css": "position: relative",
    "shortTokens": [
      "relative",
    ],
    "tokens": [
      "relative",
    ],
  },
  {
    "css": "top: 10px",
    "shortTokens": [
      "t-2.5",
    ],
    "tokens": [
      "top-2.5",
    ],
  },
  {
    "css": "left: 20px",
    "shortTokens": [
      "l-5",
    ],
    "tokens": [
      "left-5",
    ],
  },
  {
    "css": "right: 30px",
    "shortTokens": [
      "r-7.5",
    ],
    "tokens": [
      "right-7.5",
    ],
  },
  {
    "css": "bottom: 40px",
    "shortTokens": [
      "b-10",
    ],
    "tokens": [
      "bottom-10",
    ],
  },
]
`;

exports[`analyzePared > transfrom parsed 2`] = `[]`;
